### PR TITLE
Work around for null tree paths in finally blocks

### PR DESCRIPTION
**Pull request by smillst**
This is a work around for issue #1520 and #1546.   It still isn't clear why `AnnotatedTypeFactory#getPath` can't find the tree path in these cases.

### CODE REVIEW

File: checker/src/org/checkerframework/checker/nullness/KeyForAnnotatedTypeFactory.java [hunk 1]
```diff
@@ -95,7 +96,11 @@ public KeyForAnnotatedTypeFactory(BaseTypeChecker checker) {
         Pair<Tree, AnnotatedTypeMirror> context = getVisitorState().getAssignmentContext();
 
         if (returnType.getKind() == TypeKind.DECLARED && context != null && context.first != null) {
-            AnnotatedTypeMirror assignedTo = TypeArgInferenceUtil.assignedTo(this, getPath(tree));
+            TreePath path = getPath(tree);
```
Comment (wmdietl) [inline checker/src/org/checkerframework/checker/nullness/KeyForAnnotatedTypeFactory.java]: How about putting this logic, which is duplicated in KeyForPropagationTreeAnnotator, into TypeArgInferenceUtil.assignedTo? If the path is null that method can just return null and then here things would behave the same.

File: framework/src/org/checkerframework/framework/flow/CFAbstractTransfer.java [hunk 2]
```diff
@@ -171,6 +171,10 @@ protected V getValueFromFactory(Tree tree, Node node) {
         Tree preTree = analysis.getCurrentTree();
         Pair<Tree, AnnotatedTypeMirror> preCtxt = factory.getVisitorState().getAssignmentContext();
         analysis.setCurrentTree(tree);
+        TreePath previousTreePath = factory.getVisitorState().getPath();
+        if (node instanceof MethodInvocationNode) {
```
Comment (wmdietl) [inline framework/src/org/checkerframework/framework/flow/CFAbstractTransfer.java]: Can you add a comment why this is only needed for MethodInvocationNode? The assignment context also seems to be set for field updates; should the path also be set?

Comment (smillst) [reply framework/src/org/checkerframework/framework/flow/CFAbstractTransfer.java]: (Comment added below this line.)

File: checker/src/org/checkerframework/checker/nullness/KeyForPropagator.java [hunk 3]
```diff
@@ -134,6 +138,31 @@ public void propagate(
         }
     }
 
+    public void propagateNewClassTree(
```
Comment (wmdietl) [inline checker/src/org/checkerframework/checker/nullness/KeyForPropagator.java]: Should this just be default visibility?
Some javadoc might still be nice.

File: framework/src/org/checkerframework/framework/flow/CFAbstractTransfer.java (hunk 2 above)
Comment (wmdietl) [reply framework/src/org/checkerframework/framework/flow/CFAbstractTransfer.java]: @bitterfox When you look at #1520, can you undo this change?

**Pull request merged**